### ISSUE TITLE
Building variable fonts from non-default layer sources

### ISSUE THREAD

**Issue by verbosus** [googlefonts/fontmake#983]
Following up from #980 and [this comment](https://github.com/googlefonts/fontmake/pull/980#issuecomment-1438259929) from @anthrotype. The current code works for bulding static, interpolated instances from masters that are on a non-default layer, but fails to build variable fonts of the same kind. The first error I encountered is this: 
```
fontmake: Error: In 'sources/testfamily.designspace': Generating fonts from Designspace failed: "'name' table not found"
```
That’s an error that comes from `ufo2ft` and I believe it is due to a wrong assumption of considering non-default layers as being meant for sparse master support. Layers, as defined in the UFO spec, don’t have these semantics; they are meant to be generic and used for anything, including the use case we’re trying to support here. 

So when `ufo2ft` sees a master that has a `layerName` it actually strips a bunch of tables from the intermediate TTFont it’s generating internally. The tables that are allowed to pass through are defined by the [`SPARSE_TTF_MASTER_TABLES`](https://github.com/googlefonts/ufo2ft/blob/543874e2c78579905c97fff3808d71b1418d07d3/Lib/ufo2ft/constants.py#L3) constant (and likely its *OTF variant, right below.) 

I tried adding the missing `cmap`, `OS/2`, `hhea`, and `name` tables to the constant, and the error goes away. Indeed, the whole `ufo2ft` and `fontmake` test suites pass with these back on, so I’m not clear why they are being stripped in the first place. Perhaps @anthrotype can comment, since he seems to be the author of the original list. 

Anyhow, after that is fixed, another issue surfaces. While building the variable font, right after the master fonts are loaded and before starting to add various variation tables, the `varLib` model seems to get confused and, for a family with three masters on a width axis, returns locations that look like this: 
```
[{}, {'wdth': -1.0}, {'wdth': 1.0}]
```
In other words, the default location goes missing and `varLib` errors out with:
```
fontTools.varLib.errors.VariationModelError: Base master not found.
```
So the bug appears to be inside `fontTools.varLib` rather than `fontmake`, although I guess we can track it here. 

I am not entirely sure what is happening, though, and am a bit in the dark on `varLib` internals so will need some more time to dig this out.

**Comment by verbosus** [comment]
OK, so reading more of the code, it seems pretty clear that, actually, the `{}` _is_ the default location, and contrary to what I wrote earlier it doesn’t go missing, it actually gets identified. Cool. So that still doesn’t explain why the code would error out saying the base master has not been found.

**Comment by verbosus** [comment]
I think I found the culprit. The “Base master not found” exception is triggered when attempting to build an `MVAR` entry for the `underlinePosition` property, which is in the `post` table. But that is special-cased both by `varLib` and by `ufo2ft` to be a sentinel value of -0x8000. [Here’s the `ufo2ft` bit of the code](https://github.com/googlefonts/ufo2ft/blob/00a9627b8bf5d823c2202b2c6be888dffaab7d50/Lib/ufo2ft/__init__.py#L328-L337) that writes the sentinel and [here is the correspoding `varLib` reading code](https://github.com/fonttools/fonttools/blob/1a202ab1bebe9a70f77300586905351033538806/Lib/fontTools/varLib/__init__.py#L610-L617) which is marked as “HACK” in its own comment. I’m reposting the comments to both pieces of code here, for expediency. First, `ufo2ft`:
```
for sparse masters (i.e. containing only a subset of the glyphs), we
need to include the post table in order to store glyph names, so that
fontTools.varLib can interpolate glyphs with same name across masters.
However we want to prevent the underlinePosition/underlineThickness
fields in such sparse masters to be included when computing the deltas
for the MVAR table. Thus, we set them to this unlikely, limit value
(-36768) which is a signal varLib should ignore them when building MVAR.
```
And here’s `varLib`:
```
HACK: we need to special-case post.underlineThickness and .underlinePosition
and unilaterally/arbitrarily define a sentinel value to distinguish the case
when a post table is present in a given master simply because that's where
the glyph names in TrueType must be stored, but the underline values are not
meant to be used for building MVAR's deltas. The value of -0x8000 (-36768)
the minimum FWord (int16) value, was chosen for its unlikelyhood to appear
in real-world underline position/thickness values.
```
I think the assumption that any designspace source with `layerName` must therefore be a sparse master is incorrect as per the UFO and designspaceLib specs, and so is the assumption that such a source should not contribute `post` entries. I guess the `varLib` reading code relying on a sentinel value is fair enough, the semantic issue appears to be is in the `ufo2ft` writing code. Commenting it out makes a bunch of tests fail, because it breaks the way `ufo2ft` currently supports sparse masters, so it will need some heavier-handed surgery.

**Comment by verbosus** [comment]
I looked in the `ufo2ft` test code, and there isn’t anything that references specific `post` values, so I’m not clear why the sentinel-setting code is even there in the first place. Indeed, commenting out the sentinel-setting, and amending a few entries in three `MVAR` XML dumps makes both the `ufo2ft` test suite and the `fontmake` pass. @anthrotype I need some clarification on the the whole sparse master implementation across `fontmake`, `ufo2ft` and `fontTools`.

**Comment by anthrotype** [comment]
Sorry I don't have time right now to go too deep into this

> a wrong assumption of considering non-default layers as being meant for sparse master support.

you may argue it's wrong to assume that just because a layerName is not None, then ds=>ufo=>VF build pipeline assumes that only that layer's glyphs are going to contribute to the designspace and not the rest of info/kerning.
Right now that's how things work. To break that assumption would require adding additional metadata to mark a given DS source as contributing only the glyph outlines or also other font global stuff. If you're thinking of using the mutatorMath's `muteKerning` or `muteInfo` DS attrinbutes that becomes tricky because as you move from UFOs to TTFs (i.e. what varLib sees as masters) it's no longer clear what "info" or "kerning" mean in that context.

**Comment by anthrotype** [comment]
even in your other PR, you only unlocked the possibility of using the non-default layer as the base/neutral master when instantiating UFOs in fontmake.instatiator; only in those circumstances, the UFO with layerName not None contributes font-wide data like info/kerning; when the other masters use layeName, they only contribute glyphs.

**Comment by verbosus** [comment]
> To break that assumption would require adding additional metadata to mark a given DS source as contributing only the glyph outlines or also other font global stuff. If you're thinking of using the mutatorMath's `muteKerning` or `muteInfo` DS attrinbutes that becomes tricky because as you move from UFOs to TTFs (i.e. what varLib sees as masters) it's no longer clear what "info" or "kerning" mean in that context.

Since `source` elements don’t have their own lib in the DS file because they defer to the the individual master, perhaps we could include that information in the master UFO lib. Perhaps a key like `com.github.googlei18n.ufo2ft.isSparseMaster` set to `True` could trigger the current `ufo2ft` semantics, and its absence or it being set to `False` would make that master contribute global stuff. What do you think?

**Comment by verbosus** [comment]
Or we could reverse that logic, leave the current behavior as-is and bypass it with some other `com.github.googlei18n.ufo2ft.` namespaced key (naming suggestions welcome.) It seems a bit counterintuitive and overly complex, but I guess it would be the safer option, compatibility-wise.

**Comment by verbosus** [comment]
I don’t quite like any of these options, but at least they’re a start:

- `com.github.googlei18n.ufo2ft.isFullMaster` (as opposed to `isSparseMaster` which would be the current default semantics)
- `com.github.googlei18n.ufo2ft.isNotSparseMaster`
- `com.github.googlei18n.ufo2ft.masterContributesGlobalData` (as opposed to not contributing)
